src/features/device-state/state-get-utils.ts

```diff
@@ -65,10 +72,42 @@ export const getConfig = (
 	return config;
 };
 
-export const serviceInstallFromImage = (
+type ExpandedService = Pick<Service['Read'], 'id' | 'service_name'> & {
+	service_environment_variable: EnvVarList;
+	service_label: Array<{ label_name: string; value: string }>;
+};
+
+type ExpandedServiceInstall = {
+	id: ServiceInstall['Read']['id'];
+	service: ExpandedService[];
+	device_service_environment_variable: EnvVarList;
+};
+
+export function serviceInstallFromImage(
+	deviceOrFleet: { service_install: ExpandedServiceInstall[] },
+	image?: {
+		is_a_build_of__service: number | { __id: number };
+	},
+): ExpandedServiceInstall | undefined;
+export function serviceInstallFromImage(
+	deviceOrFleet: { service: ExpandedService[] },
+	image?: {
+		is_a_build_of__service: number | { __id: number };
+	},
+): ExpandedService | undefined;
+/** @deprecated Use the strongly-typed overloads with `{ service_install }` or `{ service }` instead */
```

Comment:
Is this actually still needed now that the overloads cover more/different cases?

Reply:
It is if we want to keep runtime code (in state v3) as is. The reason is in short this line: https://github.com/balena-io/open-balena-api/blob/master/src/features/device-state/routes/state-get-v3.ts#L108

Even if we overload this function so it knows that when device is undefined application will have the a correct "service" typescript is not able to to infer the proper type with a ?? so this line would have to become something like

```typescript
const si = device != null ? serviceInstallFromImage(device, image) : serviceInstallFromImage(application, image);
```

Or even a better approach for narrower type would be to on top level separate a path where device is null and one where it isn't, however, I am slightly scared about how this would affect performance (but would for sure allow better typings). Right now what I was trying to do was bend the typings to the current code instead of the code to the force correct typings.

Reply:
So I would say at least for now we should keep it and then maybe we can improve when we are actually touching state-get-V3.